kgraphql-ktor/src/main/kotlin/com/apurebase/kgraphql/KtorFeature.kt

```diff
@@ -1,4 +1,34 @@
 package com.apurebase.kgraphql
 
-// TODO: Implement
-fun ktorFeature() = println("Hello World")
+import com.apurebase.kgraphql.schema.dsl.SchemaBuilder
+import io.ktor.response.respond
+import io.ktor.routing.Route
+import io.ktor.routing.post
+import io.ktor.application.*
+import io.ktor.features.ContentNegotiation
+import io.ktor.http.ContentType
+import io.ktor.request.receive
+import io.ktor.serialization.serialization
+import io.ktor.util.KtorExperimentalAPI
+import kotlinx.serialization.UnstableDefault
+import kotlinx.serialization.json.Json
+
+const val GRAPHQL_ENDPOINT = "graphql"
+
+@KtorExperimentalAPI
+@UnstableDefault
+fun Route.graphql(ctx: Context? = null, block: SchemaBuilder.() -> Unit) {
+    val schema = KGraphQL.schema(block)
+    install(ContentNegotiation) {
+        register(ContentType.Application.Json, GraphqlSerializationConverter())
+        serialization(json = Json.nonstrict)
+    }
+    post(GRAPHQL_ENDPOINT) {
+        val request = call.receive<GraphqlRequest>()
+        val result = if (ctx != null)
+            schema.execute(request.query, request.variables.toString(), ctx)
+        else
+            schema.execute(request.query, request.variables.toString())
+        call.respond(result)
+    }
+}
```

Comment:
The context needs to be set per request.

It could maybe be done in a way like this:

```kotlin
fun Route.graphql(ctxBuilder: ContextBuilder.(ApplicationCall) -> Unit = {}, block: SchemaBuilder.() -> Unit) = route(GRAPHQL_ENDPOINT) {
    val schema = KGraphQL.schema(block)

    install(ContentNegotiation) {
        register(ContentType.Application.Json, GraphqlSerializationConverter())
        serialization(json = Json.nonstrict)
    }

    post {
        val request = call.receive<GraphqlRequest>()

        val ctx = context { ctxBuilder(this, call) }

        val result = schema.execute(request.query, request.variables.toString(), ctx)
        call.respond(result)
    }
}
```


then  in the Application.kt example you would do something like this:
```kotlin
...
install(Authentication) {
    basic {
        realm = "ktor"
        validate {
            User(4, it.name)
        }
    }
}
routing {
    authenticate(optional = true) {
        graphql(contextSetup) { ... }
    }
}

...
data class User(val id: Int = -1, val name: String = ""): Principal
private val contextSetup: ContextBuilder.(ApplicationCall) -> Unit = { call ->
    call.authentication.principal<User>()?.let {
        +it
    }
}


...
// Then this resolver should work
query("user") {
    description = "Returns a single User based on the id or the authenticated user."

    resolver { ctx: Context, id: Int? ->
        if (id == null) {
            User(4, ctx.get<User>()?.name ?: throw NotAuthenticatedException())
        } else {
            users.getOrNull(id - 1) ?: throw NotFoundException("User with id: $id does not exist")
        }
    }
}
```

A request like this:
```http
POST http://localhost:8080/graphql
Content-Type: application/json
Authorization: basic amVnZ3k6dGVzdA==

{
  "operationName": null,
  "variables": {},
  "query": "{ user { id, name } } "
}
```
will now return:
```json
{
  "data" : {
    "user" : {
      "id" : 4,
      "name" : "jeggy"
    }
  }
}
```

Reply:
Done 👍